content/recipes/authoring/kata-snippets/full-tests.md

```diff
@@ -5,6 +5,79 @@ prev: /recipes/authoring/kata-snippets/sample-tests/
+- **Make debugging of random tests easy.** Sometimes, when random tests fail, they are very difficult to debug because input is very large and cannot be easily printed. If necessary, split your random tests into two batches: one with small, debuggable inputs, and the other one with proper, large input values. Note that both parts should still contain all applicable scenarios.
+
+## Performance Tests
+
+Some kata require solutions to be fast, or to fall into some performance category. For example, the author may only wish to accept solutions completing in (sub-)linear time. Building such test suites is not an easy task!
+
+- **Performance tests can be implemented in terms of random tests.** Usually, it's not reqired to have separate sections for random tests and performance tests. Testing of performance can be driven by size of random inputs. But still, to make debugging easier, it can be worth in some cases to have a set of separate random tests with small inputs.
+- If the difficulty of a kata is roughly proportional to the size of the input, it's usually **better to have a few tests with large input rather than many with small inputs.** For example, 200 tests with huge numbers or arrays is usually better than 10,000 tests with small ones (but see remarks on size of inputs above).
+- **The difference between accepted and rejected solutions should be easy to spot.** Ideally, accepted solutions should complete well under the time limit, while rejected solutions should time out consistently. Otherwise, you risk that solutions with valid complexity characteristics will time out, and users will be frustrated looking for micro-optimizations. This approach usually calls for very large inputs, so be careful!
+- **Performance tests should be consistent between runs.** It should not happen that one and the same solution sometimes passes, and sometimes fails, depending on randomized inputs.
+- When you are going to reject solutions based on their performance, **make sure that what you measure is what you want**. It's possible that most of the available time will be consumed not by user solution, but, for example, by your input generators, or reference solution. Make sure that poor performance of your test suite does not affect the final decision whether the submitted solution is performant enough or not!
+
+## Tests with Additional Restrictions
+
+Some kata may impose additional restrictions on submitted solutions unrelated to its extensional behavior. For example, code golf kata usually require submitted solutions to be shorter than X characters, while others may require the task to be realized without a certain library or language feature. Every such requirement must be tested explicitly, but authors should be aware that it is difficult (or even infeasible) to do so _in a proper manner_:
+
+- If the user solution should be inspected as source code, it can be found in `/home/codewarrior/solution.txt`.
```

Comment:
```suggestion
- If the user solution should be inspected as source code, it can be found in `/workspace/solution.txt`.
```
`/home/codewarrior/solution.txt` has been deprecated for a while. The old language versions have `/home/codewarrior` symlinked to `/workspace` so `/worspace/solution.txt` works with any language and version.